mailbox/plugin/quota-mailing-memory/src/main/java/org/apache/james/eventsource/InMemoryEventStore.java

```diff
@@ -76,42 +80,81 @@ public void appendAll(List<? extends Event> events) {
         if (events.isEmpty()) {
             return;
         }
-        Preconditions.checkArgument(belongsToSameAggregate(events));
-        AggregateId aggregateId = events.stream().map(Event::getAggregateId).findFirst().get();
-        Optional<EventId> expectedHistoryVersion = events.stream().map(Event::eventId).min(Comparator.naturalOrder()).get().previous();
-        EventId updatedHistoryVersion = events.stream().map(Event::eventId).max(Comparator.naturalOrder()).get();
+        AggregateId aggregateId = getAggregateId(events);
+        Optional<EventId> expectedHistoryVersion = computeExpectedHistoryVersion(events);
+        EventId updatedHistoryVersion = computeUpdatedHistoryVersion(events);
 
         if (!expectedHistoryVersion.isPresent()) {
-            InMemoryHistory inMemoryHistory = store.putIfAbsent(aggregateId, InMemoryHistory.of(History.of(updatedHistoryVersion, events)));
-            if (inMemoryHistory != null) {
-                throw new EventBus.EventStoreFailedException();
-            }
-            return;
+            appendToNonExistingHistory(aggregateId, events, updatedHistoryVersion);
+        } else {
+            appendToExistingHistory(aggregateId, events, expectedHistoryVersion.get(), updatedHistoryVersion);
         }
+    }
 
-        InMemoryHistory expectedHistory = InMemoryHistory.of(History.of(expectedHistoryVersion.get(), ImmutableList.of()));
-        List<Event> updatedEvents = updatedEvents(events, aggregateId);
-        InMemoryHistory updatedHistory = InMemoryHistory.of(History.of(updatedHistoryVersion, updatedEvents));
+    public EventId computeUpdatedHistoryVersion(List<? extends Event> events) {
+        return events.stream()
+            .map(Event::eventId)
+            .max(Comparator.naturalOrder())
+            .get();
+    }
 
-        if (!store.replace(aggregateId, expectedHistory, updatedHistory)) {
+    public Optional<EventId> computeExpectedHistoryVersion(List<? extends Event> events) {
+        return events.stream()
+            .map(Event::eventId)
+            .min(Comparator.naturalOrder())
+            .get()
```

Comment:
only the caller knows the list is non-empty and can call get

Reply:
We need to add a precondition then ;-)